catalyst_voices_packages/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart

```diff
@@ -0,0 +1,358 @@
+    required this.config,
+    required this.inputs,
+    this.outputs = const [],
+    this.fee,
+    this.ttl,
+    this.auxiliaryData,
+    this.networkId,
+    this.witnessBuilder = const TransactionWitnessSetBuilder(
+      vkeys: {},
+      vkeysCount: 1,
+    ),
+  });
+
+  /// Returns a copy of this [TransactionBuilder] with extra [address] used
+  /// to spend any remaining change from the transaction, if it is needed.
+  ///
+  /// Since in a Cardano transaction the input amount must match the output
+  /// amount plus fee, the method must ensure that there are no unspent utxos.
+  ///
+  /// The algorithm first tries to create a [TransactionOutput] which will
+  /// transfer any remaining [Coin] back to the [address]. The [address]
+  /// should be the change address of the wallet initiating the transaction.
+  ///
+  /// If creating an extra [TransactionOutput] is not possible because
+  /// i.e. the remaining change is too small to cover for extra fee that such
+  /// extra output would generate then the transaction fee is increased to burn
+  /// any remaining change.
+  TransactionBuilder withChangeAddressIfNeeded(ShelleyAddress address) {
+    if (this.fee != null) {
+      // generating the change output involves changing the fee
+      return this;
+    }
+
+    final fee = minFee();
+
+    final inputTotal =
+        inputs.map((e) => e.output.amount).reduce((a, b) => a + b);
+    final outputTotal = outputs.map((e) => e.amount).reduce((a, b) => a + b);
+    final outputTotalPlusFee = outputTotal + fee;
+
+    if (outputTotalPlusFee == inputTotal) {
+      // ignore: avoid_returning_this
+      return this;
+    } else if (outputTotalPlusFee > inputTotal) {
+      throw InsufficientUtxoBalanceException(
+        actualAmount: inputTotal,
+        requiredAmount: outputTotalPlusFee,
+      );
+    } else {
+      final changeEstimator = inputTotal - outputTotal;
+
+      final minAda = TransactionOutputBuilder.minimumAdaForOutput(
+        TransactionOutput(address: address, amount: changeEstimator),
+        config.coinsPerUtxoByte,
+      );
+
+      switch (changeEstimator >= minAda) {
+        case false:
+          // burn remaining change as fee
+          return withFee(changeEstimator);
+        case true:
+          final feeForChange = TransactionOutputBuilder.feeForOutput(
+            this,
+            TransactionOutput(
+              address: address,
+              amount: changeEstimator,
+            ),
+          );
+
+          final newFee = fee + feeForChange;
+
+          switch (changeEstimator >= minAda) {
+            case false:
+              // burn remaining change as fee
+              return withFee(changeEstimator);
+            case true:
+              return withFee(newFee).withOutput(
+                TransactionOutput(
+                  address: address,
+                  amount: changeEstimator - newFee,
+                ),
+              );
+          }
+      }
+    }
+  }
```

Comment:
Is it possible to break down this function into smaller functions?

Reply:
We can, I decided to keep as close to original implementation as reasonably possible: https://github.com/dcSpark/cardano-multiplatform-lib/blob/255500b3c683618849fb38107896170ea09c95dc/chain/rust/src/builders/tx_builder.rs#L1553

The more we change it the more difficult is to apply the patches if they make them. Let me know if should refactor this code with regards to above

Reply:
That rust function is huge 🤯
What do you think about adding an internal comment explaining why we decided to design the function in such a way?

Reply:
I've added this sentence:

```dart
/// Follows code style of Cardano Multiplatform Lib to make patching easy.
```